### DISCUSSION TITLE
Using inspection api to get db roundtrips in a session

### DISCUSSION THREAD

**Discussion by Youssefares** [sqlalchemy/sqlalchemy#8938] [Usage Questions] [answered]
I am interested in inspecting the number of db round trips for a code path that uses both sqlalchemy orm and table interfaces to detect regressions/silly orm usage for a web api. There isn't a ton of docs on the Runtime Inspection API, so I'd appreciate any input to point me in the right direction.

**Comment by zzzeek** [comment, marked as answer]
the most expedient way to measure number of DB round trips is to use engine [echo](https://docs.sqlalchemy.org/en/14/core/engines.html#configuring-logging) and/or logging and just observe the SQL round trips.   more automated, you can use [before_cursor_execute](https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.ConnectionEvents.before_cursor_execute) to write custom code which can intercept all SQL and update counters and/or stats.

**Reply by Youssefares** [reply]
`before_cursor_execute` seems to be called with each individual sql statement rather than a group of queries that are batched together in a single db roundtrip. Is there another event to listen for, that will be more akin to the db roundtrips, # of queries is helpful, but it'd be helpful to see how sqlalchemy batches them into db calls.

The `before_flush` and `after_flush` events seem like good candidates, but am I right in assuming these are only orm related?

**Reply by CaselIT** [reply]
> `before_cursor_execute` seems to be called with each individual sql statement rather than a group of queries that are batched together in a single db roundtrip.

I'm not sure what you mean here. There is no "batch together queries in a single db round-trip" that I'm aware of. Each statement is executed to the db one at a time. Can you clarify what you mean ?

**Reply by Youssefares** [reply]
That’s interesting. I approached this with a false assumption. So I know that if you use the orm and `add` multiple objects of the same type, sqlalchemy will batch them into a single insert. I always thought that the engine will also batch different queries into a single db call in some cases. Thank you for clarifying that. That’s really helpful I am curious why that is not the case? Is it simply not achievable i.e. Is it to honor the order of execution in case any query violates the db data integrity/any other possible errors?

**Reply by CaselIT** [reply]
Well, I'm not sure how you can batch multiple execute calls, considering that every one returns a result.
Also it seems very hard and brittle to modify user statements.

**Reply by zzzeek** [reply]
> That’s interesting. I approached this with a false assumption. So I know that if you use the orm and `add` multiple objects of the same type, sqlalchemy will batch them into a single insert. 

that is often the case, and when it does that, you will get a single event in before_cursor_execute().   The before_cursor_execute() event corresponds to an individual statement being sent to either cursor.execute() or cursor.executemany().    


> I always thought that the engine will also batch different queries into a single db call in some cases. 

you are probably thinking of cursor.executemany().   You will get one before_cursor_execute() event for that.

**Reply by Youssefares** [reply]
Thank you for clarifying, looks like I can use before_cursor_execute with confidence it maps 1:1 to a db roundtrip. 
> you are probably thinking of cursor.executemany(). You will get one before_cursor_execute() event for that.

I am curious about this, is there any available documentation, or part of code to read even on to understand/control which kind of queries go into a single executemany. I am also curious how common it is.  For ex. if I do a number of inserts on 10+ tables in an autoflush session with the orm interface, with no reads, how likely/unlikely is it that I will end up making a db call per query?

**Reply by zzzeek** [reply]
in sqlalchemy 2.0 using all databases included except MySQL, all inserts for a set of entities of a particular type that don't have references to each other should be run within an INSERT that is either a single INSERT with many VALUES added to it (this is a new feature known as "insertmanyvalues"), or will go in as a single executemany.   For "insertmanyvalues", large numbers of entries are batched, so if the batch size is 1000 and you have 10000 rows, you would get 10 INSERT statements.   insertmanyvalues is documented at https://docs.sqlalchemy.org/en/20/core/connections.html#insert-many-values-behavior-for-insert-statements .

For previous SQLAlchemy versions, the deciding factor between if statements are run one at a time or in a "batch" is if the rows have the primary key value present in the data going in.  if they dont, the ORM needs those PK values, so it must use cursor.lastrowid or RETURNING to get it back; in SQLAlhcemy 1.4, all database drivers other than psycopg2 require that we use cursor.execute() for that  (the psycopg2 driver in SQLA 1.4 uses an "insertmanyvalues" style approach).

if you have a self-referential mapping, like a graph structure, or an inheritance mapping that has relationships between different subclasses to each other, INSERT statements necessarily need to be broken up into groups so that rows which refer to each other can be INSERTed so that the dependee row goes in, giving the ORM a chance to get server-generated PK values back, before the dependent one goes in referring to those new PK values.
